### DISCUSSION TITLE
How to use the 3rd-party plugins with the standalone api?

### DISCUSSION THREAD

**Discussion by nop990** [prettier/prettier#16539] [Q&A] [answered]
I'm writing a web extension that uses highlight.js (was using Prism previously) and Prettier to format code on a webpage. Since I am using standalone Prettier, I have to add the plugins to the Prettier.format() call directly. The included plugins (typescript, estree, babel, etc.) import fine with require() statements, but I cannot figure out how to do this for the prettier-java-plugin. I posted an Issue on their github page, but haven't received a response as of yet so figured I'd try my luck here. More info/code samples on my [stack-overflow post](https://stackoverflow.com/questions/78819054/how-to-use-3rd-party-plugins-with-standalone-prettier-api)

**Comment by Poliklot** [comment, marked as answer]
Hi, @nop990!

I made a minimal working example here:

https://github.com/Poliklot/prettier-plugin-java-standalone-browser-example

The main issue is that `prettier/standalone` does not auto-load plugins from
`node_modules`. The Java plugin has to be imported and passed directly to
`plugins`.

The important part is:

```js
import * as prettier from "prettier/standalone";
import prettierPluginJava from "prettier-plugin-java";

const formatted = await prettier.format(source, {
  parser: "java",
  plugins: [prettierPluginJava],
});
```

I used `esbuild` instead of Browserify. From a fresh clone this works:

```sh
npm install
npm run start:csp-like
```

So the Prettier standalone API itself works with `prettier-plugin-java`. The
problem seems to be mostly in how the plugin and its dependencies are bundled
for the browser/extension environment.
